builder/vsphere/common/step_reattach_cdrom.go

```diff
@@ -46,43 +47,51 @@ func (s *StepReattachCDRom) Run(_ context.Context, state multistep.StateBag) mul
 		return multistep.ActionHalt
 	}
 
-	// Keep the CD-ROM devices on the image without any attached media.
 	ui.Say("Reattaching CD-ROM devices...")
-	// Remove all CD-ROM devices from the image.
-	err = vm.RemoveCdroms()
-	if err != nil {
-		state.Put("error", fmt.Errorf("error removing cdrom prior to reattaching: %v", err))
-		return multistep.ActionHalt
-	}
-	// If the CD-ROM device type is SATA, find the SATA controller.
-	if s.CDRomConfig.CdromType == "sata" {
-		if _, err := vm.FindSATAController(); err == driver.ErrNoSataController {
-			ui.Say("Adding SATA controller...")
-			// If a SATA controller is not found, add it.
-			if err := vm.AddSATAController(); err != nil {
-				state.Put("error", fmt.Errorf("error adding sata controller: %v", err))
-				return multistep.ActionHalt
-			}
+	nActableCdroms := ReattachCDRom - len(s.CDRomConfig.ISOPaths)
+	if nActableCdroms < 0 {
+		err = vm.RemoveNCdroms(int(math.Abs(float64(nActableCdroms))))
+		if err != nil {
+			state.Put("error", fmt.Errorf("error removing cdrom prior to reattaching: %v", err))
+			return multistep.ActionHalt
 		}
-	}
-
-	ui.Say("Adding CD-ROM device...")
-	// Add the CD-ROM devices to the image based on the value of `reattach_cdroms`.
-	// A valid ISO path is required for this step. The media will subsequently be ejected.
```

Comment:
Hmm, well, I suppose I can put it before `nAttachableCdroms` calculation… I initially removed it because the whole point of the function is to process `reattach_cdroms` option. I.e. if you are inside this function, then you already know what this code supposed to do.

But I suppose the comment wouldn't hurt either, so I put it back.